src/types/redis_tdigest.cc

```diff
@@ -0,0 +1,425 @@
+      : meta_data_(meta_data), centroids_(centroids) {}
+  class Iterator {
+   public:
+    Iterator(std::vector<Centroid>::const_iterator&& iter, const std::vector<Centroid>& centroids)
+        : iter_(iter), centroids_(centroids) {}
+    std::unique_ptr<Iterator> Clone() const {
+      if (iter_ != centroids_.cend()) {
+        return std::make_unique<Iterator>(std::next(centroids_.cbegin(), std::distance(centroids_.cbegin(), iter_)),
+                                          centroids_);
+      }
+      return std::make_unique<Iterator>(centroids_.cend(), centroids_);
+    }
+    bool Next() {
+      if (Valid()) {
+        std::advance(iter_, 1);
+      }
+      return iter_ != centroids_.cend();
+    }
+    bool Prev() {
+      if (Valid()) {
+        if (iter_ != centroids_.cbegin()) {
+          std::advance(iter_, -1);
+        }
+      }
+      return Valid();
+    }
+    bool Valid() const { return iter_ != centroids_.cend(); }
+    StatusOr<Centroid> GetCentroid() const {
+      if (iter_ == centroids_.cend()) {
+        return {::Status::NotOK, "invalid iterator during decoding tdigest centroid"};
+      }
+      return *iter_;
+    }
+
+   private:
+    std::vector<Centroid>::const_iterator iter_;
+    const std::vector<Centroid>& centroids_;
+  };
+
+  std::unique_ptr<Iterator> Begin() { return std::make_unique<Iterator>(centroids_.cbegin(), centroids_); }
+  std::unique_ptr<Iterator> End() {
+    if (centroids_.empty()) {
+      return std::make_unique<Iterator>(centroids_.cend(), centroids_);
+    }
+    return std::make_unique<Iterator>(std::prev(centroids_.cend()), centroids_);
+  }
+  double TotalWeight() const { return static_cast<double>(meta_data_.total_weight); }
+  double Min() const { return meta_data_.minimum; }
+  double Max() const { return meta_data_.maximum; }
+  uint64_t Size() const { return meta_data_.merged_nodes; }
+
+ private:
+  const TDigestMetadata& meta_data_;
+  const std::vector<Centroid>& centroids_;
+};
+
+uint32_t constexpr kMaxElements = 1 * 1024;  // 1k doubles
+uint32_t constexpr kMaxCompression = 1000;   // limit the compression to 1k
+
+std::optional<rocksdb::Status> TDigest::Create(engine::Context& ctx, const Slice& digest_name,
+                                               const TDigestCreateOptions& options) {
+  if (options.compression > kMaxCompression) {
+    return rocksdb::Status::InvalidArgument(fmt::format("compression should be less than {}", kMaxCompression));
+  }
+
+  auto ns_key = AppendNamespacePrefix(digest_name);
+  auto capacity = options.compression * 6 + 10;
+  capacity = ((capacity < kMaxElements) ? capacity : kMaxElements);
+  TDigestMetadata metadata(options.compression, capacity);
+
+  LockGuard guard(storage_->GetLockManager(), ns_key);
+  auto status = GetMetaData(ctx, ns_key, &metadata);
+  if (status.ok()) {
+    return {};
+  }
+
+  if (!status.IsNotFound()) {
+    return status;
+  }
```

Comment:
Can we merge these two lines together?

Reply:
Hi @mapleFU ,

Thanks for your suggestion! 😊

We need to return both status and the existence of same tdigest before creation.
I have refactored the `std::optional` with an explicit flag `exists` with docstring.

Best Regards,
Edward